primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
+    /// Convert the given index to a vector of branch indices given tree height
+    /// and arity.
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize>;
+}
+
+impl<F: Into<BigUint> + Clone> ToTraversalPath for F {
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize> {
+        let mut pos: BigUint = <F as Into<BigUint>>::into(self.clone());
+        let mut ret = vec![];
+        for _i in 0..height {
+            ret.push((&pos % (arity as u64)).to_usize().unwrap());
+            pos /= arity as u64;
+        }
+        ret
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct MerkleCommitment<T: NodeValue> {
+    /// Root of a tree
+    pub root_value: T,
+    /// Height of a tree
+    pub height: usize,
+    /// Number of leaves in the tree
+    pub num_leaves: u64,
+}
+
+/// Basic functionalities for a merkle tree implementation. Abstracted as an
+/// accumulator for fixed-length array. Supports generate membership proof at a
+/// given position and verify a membership proof.
+pub trait MerkleTreeScheme: Sized {
+    /// Merkle tree element type
+    type Element: Element;
+    /// Hash algorithm used in merkle tree
+    type Digest: DigestAlgorithm<Self::Element, Self::Index, Self::NodeValue>;
+    /// Index type for this merkle tree
+    type Index: Index;
+    /// Internal and root node value
+    type NodeValue: NodeValue;
+    /// Merkle proof
+    type MembershipProof: Clone;
+    /// Batch proof
+    type BatchMembershipProof: Clone;
+
+    /// Tree arity
+    const ARITY: usize;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn from_elems(
+        height: usize,
+        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> BigUint;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> u64;
+
+    /// Return the current root value
+    fn root(&self) -> Self::NodeValue;
+
+    /// Return a merkle commitment
+    fn commitment(&self) -> MerkleCommitment<Self::NodeValue>;
+
+    /// Returns the leaf value given a position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Leaf value at the position along with a proof.
+    ///   LookupResult::EmptyLeaf if the leaf position is empty or invalid,
+    ///   LookupResult::NotInMemory if the leaf position has been forgotten.
+    fn lookup(
+        &self,
+        pos: impl Borrow<Self::Index>,
+    ) -> LookupResult<Self::Element, Self::MembershipProof, ()>;
```

Comment:
We could have `Self::NonMembershipProof` be an associated type that defaults to `()`. That way if we were to implement this trait for some accumulator that does support non-membership proofs, this function would automatically become slightly more useful when using that specific implementation.

Reply:
There's a `universal_lookup` function in `UniversalMerkleTreeScheme`. Are you suggesting to merge this two?
@alxiong WDYT

Reply:
I see. I guess there probably is some advantage to having `universal_lookup` be separate. I could imagine an implementation that supports non-membership proofs but creating them incurs some extra cost. Then `lookup` shouldn't pay that penalty if the caller doesn't care about the non-membership proof. If they want to pay the cost then they can call `universal_lookup`. So I guess keeping it as is is probably for the best